reasoner/resolution/resolver/Root.java

```diff
@@ -0,0 +1,307 @@
+
+        @Override
+        protected void receiveAnswer(Response.Answer fromDownstream, int iteration) {
+            LOG.trace("{}: received answer: {}", name(), fromDownstream);
+
+            Request toDownstream = fromDownstream.sourceRequest();
+            Request fromUpstream = fromUpstream(toDownstream);
+
+            ResolutionAnswer.Derivation derivation;
+            if (explanations()) {
+                // TODO
+                derivation = null;
+            } else {
+                derivation = null;
+            }
+
+            ConceptMap conceptMap = fromDownstream.answer().derived().withInitialFiltered();
+            assert fromUpstream.filter().isPresent();
+            AnswerState.UpstreamVars.Derived answer = fromUpstream.partialAnswer().asIdentity()
+                    .aggregateToUpstream(conceptMap, fromUpstream.filter().get());
+            ConceptMap filteredMap = answer.withInitialFiltered();
+            if (!responseProducer.hasProduced(filteredMap)) {
+                responseProducer.recordProduced(filteredMap);
+                ResolutionAnswer resolutionAnswer = new ResolutionAnswer(answer, disjunction.toString(), derivation, self(),
+                                                                         fromDownstream.answer().isInferred());
+                submitAnswer(resolutionAnswer);
+            } else {
+                tryAnswer(fromUpstream, iteration);
+            }
+        }
+
+        @Override
+        protected void receiveExhausted(Response.Fail fromDownstream, int iteration) {
+            LOG.trace("{}: received Exhausted, with iter {}: {}", name(), iteration, fromDownstream);
+            Request toDownstream = fromDownstream.sourceRequest();
+            Request fromUpstream = fromUpstream(toDownstream);
+
+            if (iteration < responseProducer.iteration()) {
+                // short circuit old iteration exhausted messages back out of the actor model
+                submitFail(iteration);
+                return;
+            }
+            responseProducer.removeDownstreamProducer(fromDownstream.sourceRequest());
+            tryAnswer(fromUpstream, iteration);
+        }
+
+        @Override
+        protected void initialiseDownstreamActors() {
+            LOG.debug("{}: initialising downstream actors", name());
+            for (grakn.core.pattern.Conjunction conjunction : disjunction.conjunctions()) {
+                downstreamResolvers.add(registry.conjunction(conjunction));
+            }
+        }
+
+        @Override
+        protected ResponseProducer responseProducerCreate(Request fromUpstream, int iteration) {
+            LOG.debug("{}: Creating a new ResponseProducer for request: {}", name(), fromUpstream);
+            assert fromUpstream.partialAnswer().isIdentity() && fromUpstream.filter() != null;
+            ResponseProducer responseProducer = new ResponseProducer(Iterators.empty(), iteration);
+            assert !downstreamResolvers.isEmpty();
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Use of identifier that is a keyword in later versions of Java (NM_FUTURE_KEYWORD_USED_AS_IDENTIFIER)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-use-of-identifier-that-is-a-keyword-in-later-versions-of-java-nm-future-keyword-used-as-identifier).
The identifier is a word that is reserved as a keyword in later versions of Java, and your code will need to be changed in order to compile it in later versions of Java.